### ISSUE TITLE
getting/calculating download speed 

### ISSUE THREAD

**Issue by alr2413** [781flyingdutchman/background_downloader#112]
Hi, 
special thanks for developing this awesome package (this task, i.e., developing download manager based on dio, was also on my flutter todo list).
one question: 
is it possible to get/calculate the download speed (or is it possible to set the updated frequency of progress changed event)? 

Also, do you have any plan to merge "config" branch to "main" branch?

**Comment by alr2413** [comment]
sure,
I'm asking for this feature because if I want to download a very large file from a download server with the limited very low download speed, it's not possible to show the user what's going on (it's server issue or network issue, proxy, etc), so I'm looking for a way to show some realtime feedback to the user like download speed (or even the downloaded bytes)

**Comment by 781flyingdutchman** [comment]
Hi, I've implemented this on branch `dlspeed` and was wondering if you could help test this on Android, iOS and Desktop?

Please change you r pubspec.yaml to:
```
background_downloader:
    git:
      url: https://github.com/781flyingdutchman/background_downloader.git
      ref: dlspeed
```

The `TaskProgressUpdate` now has extra fields `hasNetworkSpeed` and `hasTimeRemaining` which, if true, means you can read the value of `networkSpeed` (in MB/s) and `timeRemaining`, or use the String versions `networkSpeedAsString` and `timeRemainingAsString`.  On Android only you also include these in the 'running' notification by using the templates `{networkSpeed}` and `{timeRemaining}`  - just like you would use e.g. `{filename}`. 

Let me know if it works!

**Comment by alr2413** [comment]
Wow, thanks.
Right now I can only test on android device and it's working fine.

**Comment by alr2413** [comment]
I have more suggestion as well :)

1) store/keep the size of file in database: it'll help to keep track of downloaded file and for example if user pause download process and resume it later, by comparing the stored file size with the new one that the downloader event returns, we can detect if the file size gets changed on the server side and notify the user to decide reset/continue download.

2) add one parameter to store additional/extra data of download file for later usage: my suggestion is to use `metaData` attribute for this case and define `title` attribute to set optional name for downloaded file and use this one instead of `metaData` in notification.
I saw that one guy (issue https://github.com/781flyingdutchman/background_downloader/issues/89#issuecomment-1633392216 --> this image https://user-images.githubusercontent.com/64881244/253132468-9144e080-8a75-48e3-91e9-e90a354c5355.png) used `headers` attribute to store all of the additional information and for security reasons that's not a good idea to expose/send these information on http request.

3) in android, I've seen some weird behavior on notification (some times notifications are not showing up). Also, when the download is _in progress_, the user should not be able to dismiss notification by swiping left/right on notification.

4) if download fails for any reason (like slow internet connection or server closing connection), then it start downloading again from beginning and it might cause a download loop (depends on the situation). Like all download managers, it should be possible to resume the download progress from the last downloaded byte.

**Comment by 781flyingdutchman** [comment]
Hi, thanks for testing!
To your suggestions:
1. No plans - you can store the file size yourself if you want to - the downloader is not meant to become a database for tracking all kinds of things about files, as it will bloat it for the simpler use cases
2. Use the `metaData` field: you can store a JSON object there and do anything you want with that
3. Please file a separate bug if you have issues with the notifications. It's actually expected that a user can swipe away a notification but with the upcoming configuration options I believe if you set the task to run in foreground that also means you cannot dismiss the notification (not sure)
4. Yes, I agree there's an opportunity for improvement there. It's quite complicated so not sure when I'll get to that
5. No, similar to 1) the database isn't meant to be a database to keep track of lots of things. You can easily add something around it yourself, or implement another PersistentStorage object that does all that for your use case

**Comment by alr2413** [comment]
Hi, 

No.1 & No.5: I have to check the current database handling implementation to see how I can manipulate it.
No.2: I can use `metaData` as you said but how I can show a custom name for downloaded file (instead of filename) in notification if I use `metaData` as Json object?
No.3: sure, I'll test with `space` branch to see how it works in foreground mode.
No.4: I'll see if I can fix that and raise a merge request :)

thanks

**Comment by 781flyingdutchman** [comment]
For number 2 you can set a notification configuration specific to a task,
and put anything in there that you want. Thanks for your help!

**Comment by 781flyingdutchman** [comment]
Included network speed and time remaining in V7.9.0

**Comment by 781flyingdutchman** [comment]
On point 4 (resume on failure): I have started to look into that and it's an opportunity for a few other improvements, so while I appreciate your offer for help, I don't think I will be able to accept a PR on that. Please keep an eye out for other improvement opportunities that may be more suitable fro a PR.

**Comment by 781flyingdutchman** [comment]
Resume on failure is included in version 7.9.1. Please try it and let me know if you run into any issues.
